subsys/logging/log_backend_fs.c

```diff
@@ -0,0 +1,263 @@
+		int rc;
+		struct fs_file_t *f = &file;
+
+		/* Check if new data overwrites max file size.
+		 * If so, create new log file.
+		 */
+		int size = fs_tell(f);
+
+		if (size < 0) {
+			backend_state = BACKEND_FS_CORRUPTED;
+
+			return length;
+		} else if ((size + length) > CONFIG_LOG_BACKEND_FS_FILE_SIZE) {
+			rc = allocate_new_file(f);
+
+			if (rc < 0) {
+				goto on_error;
+			}
+		}
+
+		rc = fs_write(f, data, length);
+		if (rc == -ENOSPC) {
+			if (IS_ENABLED(CONFIG_LOG_BACKEND_FS_OVERWRITE)) {
+				del_oldest_log();
+
+				return 0;
+			}
+			/* If overwrite is disabled, full memory
+			 * is equivalent of corrupted backend.
+			 */
+			goto on_error;
+		} else if (rc < 0) {
+			return 0;
+		}
+
+		fs_sync(f);
+	}
+
+	return length;
+
+on_error:
+	backend_state = BACKEND_FS_CORRUPTED;
+	return length;
+}
+
+static uint8_t __aligned(4) buf[MAX_FLASH_WRITE_SIZE];
+LOG_OUTPUT_DEFINE(log_output, write_log_to_file, buf, MAX_FLASH_WRITE_SIZE);
+
+static void put(const struct log_backend *const backend,
+		struct log_msg *msg)
+{
+	log_backend_std_put(&log_output, 0, msg);
+}
+
+static void log_backend_fs_init(void)
+{
+	int rc;
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FS_AUTOSTART)) {
+		/* If autostart, mount file system automatically. */
+		struct fs_mount_t *mp = &lfs_storage_mnt;
+
+		rc = fs_mount(mp);
+		if (rc < 0) {
+			backend_state = BACKEND_FS_CORRUPTED;
+
+			return;
+		}
+	}
+
+	rc = allocate_new_file(&file);
+	backend_state = (rc ? BACKEND_FS_CORRUPTED : BACKEND_FS_OK);
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	/* In case of panic deinitialize backend. It is better to keep
+	 * current data rather than log new and risk of failure.
+	 */
+	log_backend_deactivate(backend);
+}
+
+static void dropped(const struct log_backend *const backend, uint32_t cnt)
+{
+	ARG_UNUSED(backend);
+
+	log_backend_std_dropped(&log_output, cnt);
+}
+
+static int allocate_new_file(struct fs_file_t *file)
+{
+	/* In case of no log file or current file fills up
+	 * create new log file.
+	 */
+	int rc;
+	int file_num = 0;
+	struct fs_statvfs stat;
+
+	assert(file);
+
+	if (backend_state == BACKEND_FS_NOT_INITIALIZED) {
+		/* Search for the last used log number. */
+		struct fs_dir_t dir = { 0 };
+
+		rc = fs_opendir(&dir, CONFIG_LOG_BACKEND_FS_MOUNT_POINT);
+
+		while (rc >= 0) {
+			struct fs_dirent ent = { 0 };
+
+			rc = fs_readdir(&dir, &ent);
+			if ((rc < 0) || (ent.name[0] == 0)) {
+				break;
+			}
+			if (strstr(ent.name, CONFIG_LOG_BACKEND_FS_FILE_PREFIX)
+			    != NULL) {
+				++file_ctr;
+				file_num = atoi(ent.name + LOG_PREFIX_LEN) + 1;
+			}
+		}
+
+		(void)fs_closedir(&dir);
+		if (rc < 0) {
+			goto out;
+		}
+
+		backend_state = BACKEND_FS_OK;
+	} else {
+		fs_close(file);
+		char *name = strstr(fname, CONFIG_LOG_BACKEND_FS_FILE_PREFIX);
+
+		file_num = atoi(name + LOG_PREFIX_LEN) + 1;
+	}
+
+	sprintf(fname, "%s/%s%04d", CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+		CONFIG_LOG_BACKEND_FS_FILE_PREFIX, file_num);
```

Comment:
When `file_num` exceeds 9999 it won't be a 4-digit sequence anymore.  I guess that'll work as long as you use the full `MAX_PATH_LEN` (probably 256) as the storage buffer, though that's overkill.

This should also use snprintf().